### ISSUE TITLE
AccessDeniedException for AWS KMS with Signatory v1.1.1

### ISSUE THREAD

**Issue by mbonenberger** [ecadlabs/signatory#388]
When using Signatory v1.1.1, accessing keys with Signatory on AWS KMS is no longer possible. It fails with the following error: 

```
Error: AccessDeniedException: User: arn:aws:iam::781xxxx:user/signatory is not authorized to perform: kms:GetPublicKey on resource: arn:aws:kms:eu-xxx:xx:key/xxx because no identity-based policy allows the kms:GetPublicKey action
	status code: 400, request id: 791xxx
```

We shifted AWS credentials to the systemd service file and only configure non-sensitive information in the signatory.yaml file, like this: 

```
vaults:
  aws:
    driver: awskms
    config:
      user_name: signatory
      region: eu-central-2
```

And then use environment variables in the service file, like this: 

```
[Service]
Environment=AWS_ACCESS_KEY_ID=xxx
Environment=AWS_SECRET_ACCESS_KEY=xxx
```

I also put the credentials in the signatory.yaml file to test if the keys can be accessed, but that's not the case. 

When downgrading to v1.0.0-beta3, the error disappears.

**Comment by stephengaudet** [comment]
hello @mbonenberger I'm sorry to hear you are having trouble with the upgrade.  
for what it's worth, AWS KMS keys in our test environments work fine with v1.1.1, so, there must be something specific to your configuration that is different.   thank you for the feedback, I am trying to reproduce the error.

**Comment by stephengaudet** [comment]
hello again @mbonenberger, I'm unable to reproduce the problem you are seeing... I can confirm that Signatory v1.1.1 is able to sign using AWS KMS with both supported configurations:  access key/secret in signatory.yaml, as well the documented env var.  
I have to conclude there is no defect with respect to AWS KMS. 

happy to help troubleshoot, email stephen.gaudet@ecadlabs.com if you'd like to setup a screen share meeting. 

regarding "I also put the credentials in the signatory.yaml file to test if the keys can be accessed, but that's not the case." please note that after making changes to signatory.yaml, the signatory service must be restarted to pick up the changes.. the config is only read at startup.

**Comment by stephengaudet** [comment]
looking closer at the error message you've reported: `Error: AccessDeniedException: User: arn:aws:iam::781xxxx:user/signatory is not authorized to perform: kms:GetPublicKey on resource: arn:aws:kms:eu-xxx:xx:key/xxx because no identity-based policy allows the kms:GetPublicKey action
	status code: 400`

are you sure there are no changes happening in IAM while you are troubleshooting?

**Comment by danielelisi** [comment]
Hi @mbonenberger, Stephen is correct. That is an IAM authorization error. 
```
arn:aws:iam::781xxxx:user/signatory is not authorized to perform: kms:GetPublicKey on resource: arn:aws:kms:eu-xxx:xx:key/xxx because no identity-based policy allows the kms:GetPublicKey action
```

Try granting the `kms:GetPublicKey ` policy to the `signatory` AWS user

**Comment by mbonenberger** [comment]
Thank you @stephengaudet and @danielelisi. This is the AWS policy I'm using: 

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "kms:GetPublicKey",
                "kms:DescribeKey",
                "kms:Verify",
                "kms:Sign"
            ],
            "Resource": [
                "arn:aws:kms:eu-central-2:xxx:key/85a7xxx",
                "arn:aws:kms:eu-central-2:xx:key/29c7xx"
            ],
            "Condition": {
                "IpAddress": {
                    "aws:SourceIp": "16.xx.xx.xx"
                }
            }
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": "kms:ListKeys",
            "Resource": "*",
            "Condition": {
                "IpAddress": {
                    "aws:SourceIp": "16.62.135.80"
                }
            }
        }
    ]
} 
```

So we are using two KMS keys, one for the baker and one for payouts. With 1.0.0-beta3 I had no issues accessing these keys, only with the newest release. Not sure what is missing in my configuration. So I'm happy to take your offer to troubleshoot @stephengaudet. 

Regarding the comment that  the config is only read at startup, I had copied the config file to make changes and troubleshoot and then pointed the signatory-cli to the modified config, like 

```signatory-cli -c signatory.yaml list```

**Comment by mbonenberger** [comment]
By having had a closer look at the outputs while writing the previous comment, I noticed that the key from the error message is not setup in AWS KMS: 

`arn:aws:kms:eu-central-2:xxx:key/b29xxx`

It might be a key that was previously generated and then deleted.

**Comment by stephengaudet** [comment]
thank you for the further details @mbonenberger 
I suspect, and suggest, that the behaviour you are seeing has been captured in #366 

When I filed #366, I hadn't realized that the behaviour was new to builds later than v1.0.0-beta3 

AWS KMS takes 30 days by default to actually delete a key that has been scheduled for deletion. 
Note, even if the key has not been scheduled for deletion, there would still be a problem.  Signatory user needs List and GetPublicKey rights on _every key in KMS (within the region)_ 

In your policy you've provided, I see that List is granted to *.  If you give the same to GetPublicKey, the problem will go away.

**Comment by mbonenberger** [comment]
Thank you @stephengaudet for these hints. 

I now structured the policy in the following way: 

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ListAndGetKeys",
            "Effect": "Allow",
            "Action": [
                "kms:ListKeys",
                "kms:GetPublicKey"
            ],
            "Resource": "*",
            "Condition": {
                "IpAddress": {
                    "aws:SourceIp": "16.xx.xx.xx"
                }
            }
        },
        {
            "Sid": "UseKeys",
            "Effect": "Allow",
            "Action": [
                "kms:DescribeKey",
                "kms:Verify",
                "kms:Sign"
            ],
            "Resource": [
                "arn:aws:kms:eu-central-2:xxx:key/85a7xxx",
                "arn:aws:kms:eu-central-2:xxx:key/29c7xxx"
            ],
            "Condition": {
                "IpAddress": {
                    "aws:SourceIp": "16.xx.xx.xx"
                }
            }
        }
    ]
}
```

However, this does not work as well. The error message I receive on both signatory versions is the following: 

```
INFO[0000] Initializing vault                            vault=awskms vault_name=aws
Error: UnsupportedOperationException: 
```

I'm not sure though if I correctly constructed the policy.

**Comment by stephengaudet** [comment]
I think the resource value of * needs to be expressed as a list of length 1 using square brackets 
`[ "*" ]`

Here is the policy I'm using in our github workflows
`{
	"Version": "2012-10-17",
	"Statement": [
		{
			"Sid": "VisualEditor0",
			"Effect": "Allow",
			"Action": [
				"kms:Sign"
			],
			"Resource": [
				"arn:aws:kms:us-west-2:<account>:key/<key1>",
				"arn:aws:kms:us-west-2:<account>:key/<key2>",
				"arn:aws:kms:us-west-2:<account>:key/<key3>"
			]
		},
		{
			"Sid": "VisualEditor1",
			"Effect": "Allow",
			"Action": [
				"kms:GetPublicKey",
				"kms:ListKeys"
			],
			"Resource": [
				"*"
			]
		}
	]
}`

**Comment by stephengaudet** [comment]
thank you for taking the time to give such detailed feedback @mbonenberger .   the need to grant GetPublicKey to * resource is in fact a regression defect introduced with the v1.1.0 release.  we'll be releasing a patch release v1.1.2 soon

**Comment by stephengaudet** [comment]
thanks again @mbonenberger.  v1.1.2 should resolve this issue. please let me know. kind regards

`Signatory v1.1.2 has just been released. This release is a patch release fixing 1 issue:

  https://github.com/ecadlabs/signatory/issues/388 - AWS KMS policy - Signatory errors without GetPublicKey applied to *

Download information can be found here https://github.com/ecadlabs/signatory/releases/tag/v1.1.2`

**Comment by mbonenberger** [comment]
Sorry for the late reply @stephengaudet. I was OOO. I confirm that the issue is solved now with v1.1.2!

**Comment by stephengaudet** [comment]
great news! thank you for the feedback Marc.  I hope you had an enjoyable time OOO.
